### PR TITLE
fix(cli/operator): convert logger.Fatal goroutines to errgroup (#2867)

### LINKED ISSUE #2867: cli/operator: errgroup lifecycle — converge startup Fatals into a single handler
## Background

The startup refactor (#2866) reshaped the operator-node boot path into a thin `runNode()` → `newNode()` → `node.start()` flow (all in `cli/operator/`) and converted the **synchronous** startup steps from `logger.Fatal` to returned errors, surfaced once via the single top-level `Fatal` in `cli/operator/start_node.go`.

It deliberately left the **long-lived background goroutines** launched by `node.start()` (`cli/operator/node.go`) still calling `logger.Fatal` on failure. A background failure therefore `os.Exit`s deep in a goroutine — bypassing the CLI's `node.Close()` teardown (p2p + db) and the single top-level handler.

**What already landed (this branch).** Two pieces the original issue treated as pending are now done on the integration branch:

- **Both HTTP servers already have a ctx-aware shutdown.** The metrics handler (`Handler.Start` in `observability/metrics/handler.go`, via #2866) and the SSV API server (`Server.Start` in `api/server/server.go`, via #2819) each now bind, serve in the background, and run a `<-ctx.Done()` → `httpServer.Shutdown(ctx)` watcher; each returns the bound address plus a `serveErr` channel that emits a non-graceful serve error and is closed silently on a clean stop. So the "servers are abandoned on normal shutdown" problem is **already fixed** — on `ctx` cancel both shut down gracefully. (`origin/main` still has the old blocking `Handler.Start(mux, addr)` / `Server.Run`; the new ctx-aware shape lives on the #2866 + #2819 lineage this issue builds on.)

What remains is purely the **`Fatal` coordination**: the bind paths and the serve-loop goroutines still call `logger.Fatal`, and nothing joins the long-lived goroutines into a group — so any one failure crashes the process past `Close()` instead of returning up.

> Code references: `cli/operator` symbols (and the server `Start` methods) are from #2866 / #2819, not yet on `main`, so they're given by file + symbol name rather than `main` blob links.

## Goal

Make **"`Fatal` once"** literal within `cli/operator`: every failure path reachable from `start()` — whether a **synchronous** bind error or a **background-goroutine** error — returns up through `runNode()` to the single `Fatal` in `start_node.go`, with `node.Close()` running on the way out. The long-lived goroutines join an `errgroup` bound to the startup `ctx`; the first failure cancels the group and returns the error, while a plain `ctx` cancellation returns `nil`. The HTTP servers join via their existing `serveErr` channels (their `Shutdown` watcher already unblocks them on cancel).

## The `Fatal`s to convert (all in `cli/operator`)

Six sites, all reachable from `node.start()`:

| # | Site | Where | Kind |
|---|------|-------|------|
| 1 | metrics bind | `node.go` `start()` — `"failed to start metrics server"` | **synchronous** |
| 2 | metrics serve loop | `node.go` `start()` — `"metrics server serve loop exited"` | goroutine |
| 3 | SSV API bind | `node.go` `start()` — `"failed to start API server"` | **synchronous** |
| 4 | SSV API serve loop | `node.go` `start()` — `"API server serve loop exited"` | goroutine |
| 5 | ongoing event sync | `eventsync.go` `syncContractEvents`'s `SyncOngoing` goroutine — `"failed syncing ongoing registry events"` | goroutine |

## HTTP servers — already ctx-aware; only optional DRY left

Both servers already bind-and-serve in the background with a `<-ctx.Done()` → `Shutdown` watcher and a `serveErr` channel (metrics: `Handler.Start`, #2866; SSV API: `Server.Start`, #2819 — see Background). So **no further server-package change is functionally required** to join the errgroup: just consume the `serveErr` channel (#2 / #4 above).

What's left is **optional cleanup, not a blocker**: the two `Start` methods are near-identical, so a shared helper (e.g. `serveHTTP`) would DRY them. Mind one signature tension if you attempt it — a blocking `serveHTTP(ctx, srv) error` conflicts with the current `Start(ctx) (boundAddr, <-chan error, err)` shape, which returns the `:0`-resolved bound address *before* serving (tests depend on this). Reconcile with that shape rather than reintroducing a blocking signature.

The WS server (`wsServer.Start`, started inside `operator.Node.Start` in `operator/node.go`, with its own goroutine `Fatal`) is **out of scope** — it's operator-package, not `cli/operator`.

## Proposed approach

1. `g, gctx := errgroup.WithContext(ctx)` in `start()`.
2. `g.Go(...)` each long-lived service: the metrics + SSV-API serve loops (`return <-serveErr`), the health prober, and the ongoing event sync. Convert the two synchronous bind errors (#1, #3) to `return err`.
3. Run the node as a group member too — `g.Go(func() error { return n.operatorNode.Start(gctx) })` — then `return g.Wait()`. (Don't `return n.operatorNode.Start(...)` *before* `g.Wait()`: that skips the group wait when `Start` itself errors.) Note `operatorNode.Start` blocks on `dutyScheduler.Wait()` for **operator** nodes — only exporter-standard blocks on `<-ctx.Done()` — and its clean-cancel→`nil` behavior is **owned by the `operator` package, not this issue** (see Out of scope).
4. Have `syncContractEvents` **return a start-func** for `SyncOngoing` rather than spawning it + `Fatal` internally, so `start()` can errgroup it.
5. Convert `startHealthProber` to return an error: clean `gctx.Done()` → `nil` (exit 0); persistent unhealth → non-zero error so the node still restarts (see *Health prober*). Extract the shared `probe` core used by both it and `ensureComponentsHealthy`.
6. **(Optional)** DRY the two server `Start` methods behind a shared `serveHTTP` helper (see *HTTP servers*). The ctx-aware serve + shutdown itself is already done — this step is cleanup only.

## Normal-shutdown invariant + test

Every joined goroutine MUST return `nil` on plain `ctx` cancellation (not a fatal-worthy error) — otherwise a clean SIGINT logs a spurious fatal. Add a test that cancels `ctx` and asserts `runNode()` / `start()` returns `nil`.

This is **harder than it first looks**: the existing in-process smoke tests (`cli/operator/node_test.go`) deliberately stop at `newNode()` and never enter `start()`. Driving `start()` to completion does real p2p `Setup`/`Start` *and* runs `operatorNode.Start` (the duty scheduler). `recordingP2PNetwork` already stubs the network, but `operatorNode` is a concrete `*operator.Node` with no test seam. So either:

- introduce an `operatorNode` interface seam (count this as part of this issue's work), or
- scope the test to the errgroup helper / prober / `serveHTTP` in isolation rather than full `start()`.

## Done when

- No `logger.Fatal` reachable from `cli/operator`'s `start()` — synchronous *or* in a goroutine — except `start_node.go`'s single top-level handler.
- A background-service failure (or a synchronous bind error) returns through the errgroup → `runNode()` → the single `Fatal`, with `node.Close()` running.
- Health-prober trip on persistent unhealth still terminates the node (non-zero exit → restart), now via `Close()` rather than `os.Exit`; a clean cancel exits 0.
- SIGINT / ctx-cancel shuts the metrics + SSV-API HTTP servers down gracefully (already true) and `runNode()` returns `nil`.
- Normal-shutdown test passes; gofmt / build / vet / test / lint green.

---
_Follow-up to #2866. #2819 (server graceful-shutdown) is merged — consume its `serveErr` surface rather than re-implementing it._

### PR DESCRIPTION
**Pull request by Roy-blox**
## Summary

Closes #2867.

All `logger.Fatal` calls inside background goroutines in `cli/operator` are replaced with errgroup-propagated errors. The first service failure now cancels the rest and surfaces as a return value from `start()`, which already ends at the single `logger.Fatal("could not start node", ...)` in `start_node.go`.

**Changes:**
- **`node.go` `start()`**: wrap with `errgroup.WithContext(n.ctx)`, replace all 4 Fatal goroutines (metrics bind/serve, API bind/serve) with `return err` / `g.Go(...)`, use `gctx` throughout all ctx-accepting calls, register `operatorNode.Start` in the group, return `g.Wait()`
- **`prober.go` `startHealthProber()`**: signature `→ error`; unhealthy probe now returns `fmt.Errorf(...)` instead of calling `logger.Fatal`
- **`eventsync.go` `syncContractEvents()`**: return `(*EventSyncer, func() error, error)`; ongoing-sync goroutine + Fatal removed; caller receives a start-func it `g.Go`s

## Test plan

- [ ] CI build passes
- [ ] Deploy to staging nodes, verify startup commit hash matches PR head
- [ ] Duties running at round 1 with no errors
- [ ] No `logger.Fatal` calls remaining in background goroutines in `cli/operator`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Roy-blox** [comment]
## QA Report: staging deploy ✅

| Field | Value |
|-------|-------|
| **Environment** | stage-hoodi |
| **Nodes** | 13-16 |
| **Sensor** | `ssv-stage-fix-errgroup-lifecycle-2867-r-53ee54-1` |
| **Deployed at** | 2026-06-16T12:38Z |

### 1. Commit verified
All 4 nodes started on commit `725368c9c7a73945c3deda1e246e0d6dba75cdae` (PR head).
```
2026-06-16T12:38:12Z ssv-node-13 INFO  starting SSV-Node:untagged-725368c9c7a73945c3deda1e246e0d6dba75cdae
2026-06-16T12:38:12Z ssv-node-14 INFO  starting SSV-Node:untagged-725368c9c7a73945c3deda1e246e0d6dba75cdae
2026-06-16T12:38:12Z ssv-node-15 INFO  starting SSV-Node:untagged-725368c9c7a73945c3deda1e246e0d6dba75cdae
2026-06-16T12:38:12Z ssv-node-16 INFO  starting SSV-Node:untagged-725368c9c7a73945c3deda1e246e0d6dba75cdae
```

### 2. Duty execution
Duties completing at round 1 on all nodes:
```
ssv-node-14 INFO  ✔️successfully finished duty processing runner_role=AGGREGATOR slot=3283293 consensus_rounds=1 total_consensus_time=0.062s total_duty_time=7.9s
```

### 3. Errors
Only benign P2P gossip noise (`validation ignored`) — same pattern as baseline. No application errors, no panics, no `Fatal` logs.

### 4. Metrics

| Operator | Peers | CL Sync | Consensus Avg (AGGREGATOR, 5m) |
|----------|-------|---------|-------------------------------|
| 13 | 7 | synced | 0.0065s |
| 14 | 8 | synced | 0.0077s |
| 15 | — | synced | 0.0066s |
| 16 | 7 | synced | 0.0081s |

Consensus at 6-13ms across duty types — normal for stage.

### 5. errgroup-specific checks
- No `FATAL` entries in any node's logs after startup — confirms Fatal goroutines are gone
- Nodes ran for 10+ minutes without any unexpected restarts or exits
- Health prober, ongoing event sync, metrics server, and `operatorNode.Start` all running as errgroup members with no premature cancellations observed

**Comment by Roy-blox** [comment]
**Greptile's concerns — addressed in follow-up commit `fc70bdf`:**

**1. `startHealthProber` ctx.Err() guard — valid, fixed.**

The concern is correct: if a component's `Healthy()` implementation returns a non-context error when its network connection is torn down (e.g. `io.EOF` instead of `context.Canceled`), `ProbeAll` would return that error, and the old code would propagate it through the errgroup to `start_node.go`'s `logger.Fatal("could not start node")` — making a clean shutdown look like a startup failure.

Fixed: added `if ctx.Err() != nil { return nil }` before the error return in `startHealthProber`. If the context is already cancelled when ProbeAll fails, we return nil — the errgroup already has the root cause (or it's a clean shutdown). Added a test (`Test_startHealthProber_ctxCancelMasksProbeFail`) that covers this path.

**2. `ongoingSync` `context.DeadlineExceeded` — valid, fixed.**

The original `!errors.Is(err, context.Canceled)` filter misses `context.DeadlineExceeded`. Changed to `ctx.Err() == nil` which is both shorter and complete: if the context is done for any reason (canceled or deadline exceeded), the error is a shutdown artifact and we return nil.

File: cli/operator/node.go [hunk 1]
```diff
@@ -529,15 +532,18 @@ func (n *node) start() error {
...
 	}
+	if ongoingSync != nil {
+		g.Go(ongoingSync)
+	}
 
-	go startHealthProber(n.ctx, n.logger, healthProber)
+	g.Go(func() error { return startHealthProber(gctx, n.logger, healthProber) })
 
-	if _, err := n.metadataSyncer.SyncAll(n.ctx); err != nil {
+	if _, err := n.metadataSyncer.SyncAll(gctx); err != nil {
```
Comment (momosh-ssv) [inline cli/operator/node.go:541]: What do you think about these early returns racing a backgrounded failure? 

Once `ongoingSync`/prober/metrics are `g.Go`'d, a failure cancels `gctx`, so this path returns `failed to sync metadata on startup: context canceled` while the real cause sits in the un-waited errgroup — the funnel Fatal then misattributes the crash. 

Maybe `return errors.Join(err, g.Wait())` on these early returns so the originating error wins?

File: cli/operator/node.go [hunk 2]
```diff
@@ -599,21 +605,26 @@ func (n *node) start() error {
...
-	}
-	if err := n.operatorNode.Start(n.ctx); err != nil {
-		return fmt.Errorf("failed to start SSV node: %w", err)
+			return nil
+		})
 	}
 
-	return nil
+	g.Go(func() error {
+		if err := n.operatorNode.Start(gctx); err != nil {
```
Comment (momosh-ssv) [inline cli/operator/node.go:621]: Should we consider that `operatorNode.Start` still calls `logger.Fatal` internally (WS serve loop, `operator/node.go:445`)? 

Running inside this `g.Go`, it `os.Exit`s past the errgroup and `Close()` on a mid-run WS failure — the same pattern #2867 set out to remove, just one layer down. 

Worth a follow-up, or threading that `serveErr` out like metrics/API do?

File: cli/operator/prober_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// Test_startHealthProber_ctxCancelMasksProbeFail verifies the ctx.Err() guard: when ctx is
+// canceled at the same time a probe fails (e.g. the network stack tears down components before
+// the errgroup context fully propagates), startHealthProber returns nil rather than a misleading
+// "components unhealthy" error that would cause a Fatal log on normal shutdown.
+func Test_startHealthProber_ctxCancelMasksProbeFail(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel() // cancel before starting — simulates shutdown racing a probe
+
+	prober := hprobe.NewHealthProber(zap.NewNop())
+	prober.AddComponent("cl", simpleComponent{err: errors.New("broken")}, 100*time.Millisecond, 0, 0)
```
Comment (momosh-ssv) [inline cli/operator/prober_test.go:56]: Might be worth strengthening this case — the stub's `Healthy` ignores ctx and returns the same error regardless, so the test passes via the guard but doesn't actually prove it distinguishes a real probe failure from a cancel. 

A component that respects ctx would return `context.Canceled` (swallowed to nil in `probeComponent`), so `ProbeAll` returns nil and the guard is never reached. A ctx-respecting stub would exercise the realistic teardown race the comment describes.

File: cli/operator/eventsync.go [hunk 4]
```diff
@@ -137,19 +141,18 @@ func syncContractEvents(
...
-					zap.Error(err),
-				)
+		// Return a start-func for ongoing event sync. The caller (start()) registers it with errgroup
+		// so a failure propagates as a clean error rather than a Fatal. The node can't operate without
+		// staying current with Ethereum events, so any non-canceled failure exits the process.
+		ongoingFromBlock := fromBlock.Uint64()
+		ongoingSync = func() error {
+			err := eventSyncer.SyncOngoing(ctx, ongoingFromBlock)
+			if err != nil && ctx.Err() == nil {
+				return fmt.Errorf("failed syncing ongoing registry events (from_block=%d): %w", ongoingFromBlock, err)
```
Comment (momosh-ssv) [inline cli/operator/eventsync.go:151]: Seems that the old Fatal logged `last_processed_block` while this now logs `from_block` (= `last_processed_block + 1`). 

Minor, but operators grepping for `last_processed_block` on this failure won't find it anymore — maybe include both fields?
